src/parser/open-api-v3/open-api-v3-parser.js

```diff
@@ -0,0 +1,215 @@
+import Immutable from 'immutable';
+import refParser from 'json-schema-ref-parser';
+import getUIReadySchema from './schemaParser';
+
+/**
+ * Converts openApiV3 object to new object ready to be consumed by the UI
+ *
+ * @param {Object} openApiV3
+ *
+ * @return {Object}
+ */
+export default async function getUIReadyDefinition(openApiV3) {
+  let derefOpenApiV3;
+  try {
+    derefOpenApiV3 = await refParser.dereference(openApiV3);
+  } catch (error) {
+    throw new Error(`Unable to dereference input definition. Details: ${JSON.stringify(error)}`);
+  }
+
+  const info = derefOpenApiV3.info;
+  const paths = derefOpenApiV3.paths;
+
+  // Get tags
+  const tags = getTags(paths);
+
+  // Construct navigation
+  const navigation = getUIReadyNavigation(paths, tags);
+
+  // Construct main contents
+  const services = getUIReadyServices(paths, tags);
+
+  const definition = new Immutable.Map({
+    title: info.title,
+    version: info.version,
+    description: info.description,
+    navigation,
+    services,
+  });
+
+  return definition;
+}
+
+/**
+ * Construct responses object ready to be consumed by the UI
+ *
+ * @param {Object} responses
+ *
+ * @return {Object}
+ */
+function getUIResponses(responses) {
+  const uiResponses = {};
+
+  for (const statusCode in responses) {
+    const response = responses[statusCode];
+
+    const uiResponse = {
+      code: statusCode,
+      description: response.description,
+    };
+
+    if (response.content) {
+      const responseContent = response.content['application/vnd.api+json'] || response.content['application/json'];
+      uiResponse.schema = getUIReadySchema(responseContent.schema);
+    }
+
+    uiResponses[statusCode] = uiResponse;
+  }
+
+  return uiResponses;
+}
+
+/**
+ * Construct request object ready to be consumed by the UI
+ *
+ * @param {String} description
+ * @param {Object} requestBody
+ *
+ * @return {Object}
+ */
+function getUIRequest(description, requestBody = null) {
+  const uiRequest = {
+    description,
+  };
+
+  if (requestBody) {
+    const requestContent =
+      requestBody.content['application/vnd.api+json'] ||
+      requestBody.content['application/json'];
+
+    uiRequest.schema = getUIReadySchema(requestContent.schema);
+    uiRequest.example = requestContent.examples;
+  }
+
+  return uiRequest;
+}
+
+/**
+ * Construct services Map from paths object
+ *
+ * @param {Object} paths
+ * @param {Array} tags
+ *
+ * @return {Immutable.Map}
+ */
+function getUIReadyServices(paths, tags) {
+  const services = {};
+
+  tags.forEach((tag) => {
+    let uiMethods = {};
+
+    for (const pathKey in paths) {
```

Comment:
This, and the following loop are repeated three times in this file. Could it be abstracted to make this easier/more efficient?